OpenVR Module for Godot 3.0

BastiaanOlij opened the pull request:
This is a Godot 3.0 version of #7968 based on the new AR/VR server logic. 

There is a test Godot project here to test with:
https://github.com/BastiaanOlij/godot3_test_projects/tree/master/TestGodotARVR

It looks like tracking is working properly at the moment however I'm still having problems with outputting to the HMD with the newer approach.

Note that to compile you'll need to get the latest OpenVR SDK from here:
https://github.com/ValveSoftware/openvr
And place the contents in thirdparty/openvr
The files that are in the Godot repository are incomplete due to all the binaries that are delivered.

On Windows make sure to copy the openvr.dll file from thirdpaty/openvr/bin/winxx into the folder containing your godot.exe

On Mac OSX make sure to copy the framework for openvr from thirdparty/openvr/bin/osxxx into Godot.app/Contents/frameworks

I am unsure about the steps required for Linux.

BastiaanOlij committed:
OpenVR implementation based on the AR/VR server approach

Openvr: cross platform builds and fix openvr_api.lib linking.

Apparently msvc wants scons to pass the lib via LINKFLAGS.

reduz commented:
Looks great, I will give it a check when I go back to Buenos Aires

On Apr 28, 2017 11:26 AM, "Bastiaan Olij" <notifications@github.com> wrote:

> This is a Godot 3.0 version of #7968
> <https://github.com/godotengine/godot/pull/7968> based on the new AR/VR
> server logic.
>
> There is a test Godot project here to test with:
> https://github.com/BastiaanOlij/TestGodotVR
>
> I haven't had the change to test on a Vive so I am not sure if the
> controller logic will work. Also testing with TrinusVR the world seemed to
> be upside down but this may not be a problem in the logic, but a problem
> with Trinus. I hope to test things soon on a proper Vive.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/godotengine/godot/pull/8568
> Commit Summary
>
>    - Adding support for asymmetrical camera frustums for stereoscopic
>    rendering (for HMD/2DTV)
>    - Adding base classes and structures for ARVR support
>    - OpenVR implementation based on the AR/VR server approach
>
> File Changes
>
>    - *M* core/math/camera_matrix.cpp
>    <https://github.com/godotengine/godot/pull/8568/files#diff-0> (103)
>    - *M* core/math/camera_matrix.h
>    <https://github.com/godotengine/godot/pull/8568/files#diff-1> (1)
>    - *A* core/math/frustum.cpp
>    <https://github.com/godotengine/godot/pull/8568/files#diff-2> (105)
>    - *A* core/math/frustum.h
>    <https://github.com/godotengine/godot/pull/8568/files#diff-3> (56)
>    - *M* doc/base/classes.xml
>    <https://github.com/godotengine/godot/pull/8568/files#diff-4> (91)
>    - *M* drivers/gles2/rasterizer_gles2.cpp
>    <https://github.com/godotengine/godot/pull/8568/files#diff-5> (8)
>    - *M* drivers/gles2/rasterizer_gles2.h
>    <https://github.com/godotengine/godot/pull/8568/files#diff-6> (1)
>    - *M* drivers/gles3/rasterizer_storage_gles3.cpp
>    <https://github.com/godotengine/godot/pull/8568/files#diff-7> (8)
>    - *M* drivers/gles3/rasterizer_storage_gles3.h
>    <https://github.com/godotengine/godot/pull/8568/files#diff-8> (1)
>    - *M* editor/spatial_editor_gizmos.cpp
>    <https://github.com/godotengine/godot/pull/8568/files#diff-9> (47)
>    - *A* modules/openvr/SCsub
>    <https://github.com/godotengine/godot/pull/8568/files#diff-10> (22)
>    - *A* modules/openvr/config.py
>    <https://github.com/godotengine/godot/pull/8568/files#diff-11> (8)
>    - *A* modules/openvr/openvr.cpp
>    <https://github.com/godotengine/godot/pull/8568/files#diff-12> (381)
>    - *A* modules/openvr/openvr.h
>    <https://github.com/godotengine/godot/pull/8568/files#diff-13> (84)
>    - *A* modules/openvr/register_types.cpp
>    <https://github.com/godotengine/godot/pull/8568/files#diff-14> (45)
>    - *A* modules/openvr/register_types.h
>    <https://github.com/godotengine/godot/pull/8568/files#diff-15> (32)
>    - *M* platform/iphone/rasterizer_iphone.cpp
>    <https://github.com/godotengine/godot/pull/8568/files#diff-16> (8)
>    - *M* platform/iphone/rasterizer_iphone.h
>    <https://github.com/godotengine/godot/pull/8568/files#diff-17> (1)
>    - *M* scene/3d/camera.cpp
>    <https://github.com/godotengine/godot/pull/8568/files#diff-18> (109)
>    - *M* scene/3d/camera.h
>    <https://github.com/godotengine/godot/pull/8568/files#diff-19> (18)
>    - *M* servers/SCsub
>    <https://github.com/godotengine/godot/pull/8568/files#diff-20> (1)
>    - *A* servers/arvr/SCsub
>    <https://github.com/godotengine/godot/pull/8568/files#diff-21> (7)
>    - *A* servers/arvr/arvr_interface.cpp
>    <https://github.com/godotengine/godot/pull/8568/files#diff-22> (69)
>    - *A* servers/arvr/arvr_interface.h
>    <https://github.com/godotengine/godot/pull/8568/files#diff-23> (78)
>    - *A* servers/arvr/arvr_positional_tracker.cpp
>    <https://github.com/godotengine/godot/pull/8568/files#diff-24> (264)
>    - *A* servers/arvr/arvr_positional_tracker.h
>    <https://github.com/godotengine/godot/pull/8568/files#diff-25> (101)
>    - *A* servers/arvr_server.cpp
>    <https://github.com/godotengine/godot/pull/8568/files#diff-26> (157)
>    - *A* servers/arvr_server.h
>    <https://github.com/godotengine/godot/pull/8568/files#diff-27> (95)
>    - *M* servers/register_server_types.cpp
>    <https://github.com/godotengine/godot/pull/8568/files#diff-28> (11)
>    - *M* servers/visual/rasterizer.h
>    <https://github.com/godotengine/godot/pull/8568/files#diff-29> (1)
>    - *M* servers/visual/visual_server_raster.cpp
>    <https://github.com/godotengine/godot/pull/8568/files#diff-30> (43)
>    - *M* servers/visual/visual_server_raster.h
>    <https://github.com/godotengine/godot/pull/8568/files#diff-31> (8)
>    - *M* servers/visual/visual_server_scene.cpp
>    <https://github.com/godotengine/godot/pull/8568/files#diff-32> (21)
>    - *M* servers/visual/visual_server_scene.h
>    <https://github.com/godotengine/godot/pull/8568/files#diff-33> (5)
>    - *M* servers/visual_server.h
>    <https://github.com/godotengine/godot/pull/8568/files#diff-34> (3)
>    - *M* thirdparty/README.md
>    <https://github.com/godotengine/godot/pull/8568/files#diff-35> (13)
>    - *A* thirdparty/openvr/.gitattributes
>    <https://github.com/godotengine/godot/pull/8568/files#diff-36> (42)
>    - *A* thirdparty/openvr/LICENSE
>    <https://github.com/godotengine/godot/pull/8568/files#diff-37> (27)
>    - *A* thirdparty/openvr/README.md
>    <https://github.com/godotengine/godot/pull/8568/files#diff-38> (13)
>    - *A* thirdparty/openvr/headers/openvr.h
>    <https://github.com/godotengine/godot/pull/8568/files#diff-39> (3643)
>    - *A* thirdparty/openvr/headers/openvr_api.cs
>    <https://github.com/godotengine/godot/pull/8568/files#diff-40> (4786)
>    - *A* thirdparty/openvr/headers/openvr_api.json
>    <https://github.com/godotengine/godot/pull/8568/files#diff-41> (3789)
>    - *A* thirdparty/openvr/headers/openvr_capi.h
>    <https://github.com/godotengine/godot/pull/8568/files#diff-42> (1869)
>    - *A* thirdparty/openvr/headers/openvr_driver.h
>    <https://github.com/godotengine/godot/pull/8568/files#diff-43> (2541)
>    - *A* thirdparty/openvr/lib/osx32/libopenvr_api.dylib
>    <https://github.com/godotengine/godot/pull/8568/files#diff-44> (0)
>
> Patch Links:
>
>    - https://github.com/godotengine/godot/pull/8568.patch
>    - https://github.com/godotengine/godot/pull/8568.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/8568>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AF-Z2xPqat2tZTOC_shm-LQ06dLraxujks5r0bDVgaJpZM4NLQQW>
> .
>

akien-mga commented:
Note that this PR should not be merged as is, it relies on the commits from #8567, which should thus likely be reviewed and merged before. Then this PR can be rebased and further assessed (but feel free to use the current branch for testing).

Hinsbart commented:
> the world seemed to be upside down but this may not be a problem in the logic, but a problem with Trinus. 

Same here with a vive. A simple workaround is to set flip_h on the viewport(s), but of course that doesn't fix the underlying problem...

punto- commented:
is there a dylib in the commit? if we can't distribute the code for each
implementation, they could be made as native modules

On 28 April 2017 at 09:08, Andreas Haas <notifications@github.com> wrote:

> the world seemed to be upside down but this may not be a problem in the
> logic, but a problem with Trinus.
>
> Same here with a vive. A simple workaround is to set flip_h on the
> viewport(s), but of course that doesn't fix the underlying problem...
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/8568#issuecomment-297981619>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGVmPfWPD3d_HnoAJBOaRw5AZDOD7nCUks5r0dbXgaJpZM4NLQQW>
> .
>

BastiaanOlij commented:
@akien-mga would be way to early to merge this anyway, note that this also relies on PR #7121 which has been ready for some time (which I have kept alive for Godot 2 in PR #7778). Hinsbart and I will keep working on this but it may very well be that we'll change this to a GDNative plugin and only 7121 and 8567 should eventually be merged.

@Hinsbart actually it is probably undoing a fix in the commit function. It submits the viewport texture upside down to StreamVR because in Godot 2 it was upside down. May simply no longer be required :)

@punto- yes if you look at the thirdparty folder you'll find an entry for openvr which should, but does not, contain a full copy of:
https://github.com/ValveSoftware/openvr
And yes as mentioned before we're thinking about changing the openvr implementation itself to GDNative, but as I wrote all of this originally for Godot 2.0 that is a step we have yet to make

Also I had problems setting up the scsub file correctly so anyone who can give some pointers to have this working correctly..

punto- commented:
ah I guess openvr is open source.. but there'll probably be a lot of 3rd
party stuff there, especially in the sensors, those are great candidates
for gdnative modules

On 28 April 2017 at 19:15, Bastiaan Olij <notifications@github.com> wrote:

> @akien-mga <https://github.com/akien-mga> would be way to early to merge
> this anyway, note that this also relies on PR #7121
> <https://github.com/godotengine/godot/pull/7121> which has been ready for
> some time (which I have kept alive for Godot 2 in PR #7778
> <https://github.com/godotengine/godot/pull/7778>). Hinsbart and I will
> keep working on this but it may very well be that we'll change this to a
> GDNative plugin and only 7121 and 8567 should eventually be merged.
>
> @Hinsbart <https://github.com/Hinsbart> actually it is probably undoing a
> fix in the commit function. It submits the viewport texture upside down to
> StreamVR because in Godot 2 it was upside down. May simply no longer be
> required :)
>
> @punto- <https://github.com/punto-> yes if you look at the thirdparty
> folder you'll find an entry for openvr which should, but does not, contain
> a full copy of:
> https://github.com/ValveSoftware/openvr
> And yes as mentioned before we're thinking about changing the openvr
> implementation itself to GDNative, but as I wrote all of this originally
> for Godot 2.0 that is a step we have yet to make
>
> Also I had problems setting up the scsub file correctly so anyone who can
> give some pointers to have this working correctly..
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/8568#issuecomment-298119056>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGVmPYUYg1dycQ2xUHMadFclFl6lDc81ks5r0mUGgaJpZM4NLQQW>
> .
>

BastiaanOlij commented:
@reduz , this may warrant a separate PR especially if we change the implementation over to GDNative, I've exposed the texture id of the texture object because I need to submit that to SteamVR so it can post process the texture and display it on the HMD. Unlike with the old Rift SDKs or with the Mobile version the HMD isn't a second screen anymore but fully owned by Steam VR.

This may however not be the correct way to do things. One thing that I've been wondering about is if this wouldn't be better to implement as having multiple (virtual) monitors. I don't know how Godot handles multiple monitors currently but I can imagine that instead of the commit approach we could introduce a hook that allows the Ar/Vr implementation instance to pretend each eye is a separate monitor. While in the background rendering to the monitor for an eye still renders to texture it can then submit it to the platform.

That approach would also work better for mobile VR because we can implement all the code but actually render to the main viewport and take it out of the hands of the person building the game.

BastiaanOlij commented on modules/openvr/openvr.cpp:
```diff
@@ -0,0 +1,381 @@
+			bounds.uMax = 1.0;
+			bounds.vMin = 1.0;
+			bounds.vMax = 0.0;
+
```
@Hinsbart Once I'm behind my PC (I'm on my mac) I'll change this but this is the code I was talking about, for Godot 2 I had to flip the viewport upside down, so change this to:
bounds.uMin = 0.0;
bounds.uMax = 1.0;
bounds.vMin = 0.0;
bounds.vMax = 1.0;

and it should no longer render upside down

BastiaanOlij commented:
@akien-mga as resident GIT expert. Hinsbart is going to be adding stuff he did for this as well, do we just check in our changes and then occasionally merge the commits together and rebase it or is there a better workflow when multiple people try to make a PR come together (well until/if we decide to split it as a GDNative thing)

@punto- I think open source is stretching it a bit. It's open in that it is an unrestricted API to SteamVR but all the real stuff happens inside Steam VR and inside the Steam VR drivers the venders of headsets supply which are all as closed source as it gets. 
What I do not get is why they've implemented the OpenVR as a very thin DLL that you need to supply with your application. Why isn't that DLL just part of the Steam VR installation...

levrik commented on thirdparty/README.md:
```diff
@@ -152,6 +152,19 @@ Files extracted from the upstream source:
+
+- Upstream: https://github.com/ValveSoftware/openvr
+- Version: 1.0.6
+- License: Valve (looks like MIT)
```
GitHub seems to detect `BSD 3-clause "New" or "Revised" License`, see https://github.com/ValveSoftware/openvr/blob/master/LICENSE

BastiaanOlij replied:
Yup, my bad, looks like it's BSD, I'll update it :)

BastiaanOlij commented:
I've changed the commit logic so it doesn't flip the texture upside down and changed the license info in the readme. Can't do much more at this time, I'm hoping to have access to a rift on Thursday and see if all the controller logic works.

@Hinsbart , hope I didn't cause any trouble for you as I amended my latest commit

BastiaanOlij commented:
@Hinsbart I've rebased everything to todays master and condensed the openvr things itself into a single commit. also note that there is a 1.0.7 release of openvr so you need to get new files from their github page.

I think we're at a stage now where we could implement this as an GDNative module if we build from 8567 and move the code in modules\openvr into a separate GDNative thing.

nhydock commented:
I haven't tried any builds of this yet, but how much does it impact the binary size of Godot? 
 Seeing how big of a change this is (one that I'm personally super existed for, might I add), I think it'd be nice having it available as a GDNative thing if possible, and just distribute it from the asset store and have sample projects that come with the plugin out of the box.  That way the base Godot distributable can remain light weight and focused on the basics.

BastiaanOlij commented:
@nhydock I haven't done a comparison but I don't think it increases the size by much. The real issue is the Openvr library itself which can get in the way so yes, GDNative is definitely a direction I want to start exploring

BastiaanOlij commented:
@Hinsbart , @reduz , I've rewritten this to now work on the new structure. I haven't done much testing, Godot keeps crashing on me on but I think unrelated to the changes. Not sure what is going on.

My test project https://github.com/BastiaanOlij/TestGodotVR has been updated as well though it really needs a loading window that allows us to chose the interface and give steamvr some time to start. For now make sure steamvr is up and running before you initialize the openvr interface or it will decide steamvr isn't running because its still busy starting up.

Reduz, keep in mind that you need to grab a copy of: https://github.com/ValveSoftware/openvr and place it into the thirdparty folder. Not all the files are there at the moment.
I actually have to update it to the latest version now that they have added macos support, it's currently based off of 1.0.7  or 1.7 or something.

BastiaanOlij commented:
Note, there is a new version of the test project here:
https://github.com/BastiaanOlij/godot3_test_projects/tree/master/TestGodotARVR

BastiaanOlij commented:
I've just updated this so it has been rebased to todays master and uses OpenVR 1.0.8

Still not working with my PSVR though, I think it is still having issues copying the render buffer contents to SteamVR.

BastiaanOlij commented:
Ok, I found a few bugs that were introduced in the rewrite that I totally missed. Rows and Columns are swapped for OpenVR projection matrix and somehow I had negated it (I really don't know, can't think of a scenario that would ever make sense so I must have grazed the minus key).
I had also managed to forget implementing the head tracking as we are now doing that inside of the interface instead of in game logic (originally you copied the tracker transform onto the camera node).

I'm at a point now where it is rendering correctly to the onscreen preview and tracking perfectly, but there still isn't any output to the HMD. At this point this can still be either TrinusPSVR or simply this blasted Radeon GPU in my laptop that misbehaves all the time.

I can't find anything wrong with the code other then us rendering the left eye, committing, and then rendering the right eye to the same buffer to save memory footprint. I might try and just disable one of the eyes.

BastiaanOlij commented:
@karroffel just thinking out loud here, one of the things I still want to add here is to make the models of controllers available as resources through the models API OpenVR offers (@hinsbart did an a version of this some time ago). Seeing we'll be moving this all over into a GDNative plugin we'll need to create a subclass of Mesh that has capabilities similar to PrimitiveMesh or ArrayMesh but as a ScriptMesh class or something like that?

Any suggestions here? Is there already something in the works to make mesh resources something you can implement as a GDNative module?

BastiaanOlij commented:
Changed this to work with the newer version of ARVR server

Use this branch of the test project to test:
https://github.com/BastiaanOlij/godot3_test_projects/tree/new_arvr_objects/TestGodotARVR

BastiaanOlij commented:
Okay, loads of progress today even if the changes don't show much. It's amazing what one finds out when you have access to hardware that actually works. I really need to save up for a better computer, this HP laptop is very much a let down when it comes to VR, nothing wants to work on it properly.

FInally found what had broken several weeks (months?) ago that stopped this PR from actually rendering to the HMD and it is a very simple thing. OpenVR did not like Godot's HDR render buffer. Now there has to be a setting in OpenVR somewhere to allow 16bit color channels instead of 8bit color channels as input but i haven't found it yet, why this is even important to OpenVR? I have no idea. As enabling HDR also enables some other effects it isn't impossible that the actual issue lies in one of those (MSAA maybe, there is a setting in OpenVR to allow handling that?). I will do some more experimenting with different combinations of settings to figure it out. For now however disabling HDR is the key.

Note that I possibly did find a bug here, disabling HDR in the project settings had no effect and nothing was turned on in my environment.  I ended up forcing HDR to be off by uncommenting a line in  RasterizerStorageGLES3::_render_target_allocate

The next problem to fix is the way the backgrounds are rendered when you set a background in your environment such as the horizon or panorama image. These have not been updated for stereoscopic support and well, you get sick quick :)

The final issue I ran into that I didn't realize until I tried out my space sim demo project is that there is an issue with the threads updating objects out of sync. I don't really get it because the camera is anchored to the space ship you fly around in but it started jittering. Very annoying..

Unfortunately my hands will be tied to work on this effectively for the foreseeable days because while i can borrow my mates Vive, I don't have hardware that will run it properly.. :(
I'll be working on porting my reference implementation for a GDNative based VR driver first.

beniwtv commented:
So, I had a chance to try this yesterday on Linux with the FOSS AMD drivers, and I can report I got it working, however, I had to do a few small things (sorry if I bring up issues you may have solved in your local branch already):

1. To get the linker to link against OpenVR, in the OpenVR scons file I had to change "openvr_api.lib" to just "openvr_api", else it would not find the library. Maybe there needs to be an else if case just for Linux there.
2. The bar demo project gave me an error that ArVrServer was not defined - at first thought I hadn't compiled that branch in, but it turns out that it needs to be written like "ARVRServer".
3. There's a line where the demo references ARVRServer.TRACKER_HMD, but that does not exist (I see other tracker constants tho), so I changed it to something else.
4. Controller tracking didn't seem to be working (or, at least, no controller showed up in the VR world), but that might be related to 3# and me not selecting the correct thing there. I will do more investigation there, as it did print out that it did in fact find two controllers.

So with this, I got a tracking VR headset in Godot under Linux! This is very awesome, and very promising! Great job! Can't wait for this to be in an official Godot build!

Now, while the tracking was working fine (and I couldn't see any delay with my quick testing), the lens distortion doesn't seem to be applied. In the headset the world kinda "deforms" a little while moving. Not sure if I did something wrong there.

I used SteamVR's handy "Display mirror", and it doesn't look like other applications, it seems it is just the two images rendered side by side. Usually, I would see a more rounded image there that is distorted.

Again, sorry if I brought up things you might have fixed already ;)

BastiaanOlij commented:
Hey @beniwtv , 

The issues you're having with the bar scene are probably because you checked out the master, there is a branch in that project for the latest changes. I need to merge them, i had kept those separated because originally I hadn't updated the OpenHMD and OpenVR PRs to the latest arvr server changes but that is all done now.

I don't know why the lens distortion isn't applied, that is something OpenVR should be taking care off so my best guess is that that is a driver issue on the Linux side. @Hinsbart you had this working at some point, any ideas what this could be?

beniwtv commented:
Ah, that makes sense. Will try that other branch.

It did indeed seem strange to me that the distortion is not applied - but maybe I am missing something there. I did check out OpenVR master, maybe something changed in the API?

BastiaanOlij commented:
I just merged it with the master so should be better.

OpenVR 1.0.8 should be checked in so that should be fine asfar as I know but it can't hurt updating thirdparty/openvr with fresh files from openvrs github page.. I can't test on Linux so I don't know what I may have missed or if i have outdated files..

beniwtv commented:
So, I downloaded everything again, and this time used the version of OpenVR in the repos - there was a small merge conflict too, but that was easy to solve.

With this version, lens distortion now works! The image in the headset does look normal now. So it might just be that something in OpenVR master changed. On the preview screen Godot creates on the normal PC screen when launching the scene there is now "deformation" - not sure if that is intended.

I still didn't get the controllers to track however - they're in the scene, of type ARVRController and as a child of ARVROrigin. Are they supposed to move automatically, or do they need to be controlled by a GDscript?

The OpenVR implementation itself does seem to detect both controllers - reporting two "Device X attached (vr_controller_vive_1_5)"

BastiaanOlij commented:
Hey @beniwtv ,

Good to hear it compiles now, it's probably because lib binaries aren't supposed to be added to the git repository and is in the .gitignore file, it probably never put the new versions of the binaries in. I'll see if I can force it to update the changes so that we at least get a stable repository though this will become a non issue once we port to GDNative and can include whatever we chose for that module.

The preview is probably stretched, it just blits the left eye image to screen and as it has a different aspect ratio then the lens (wider v.s higher) it stretches it out. I need to add a bit of logic so it maintains the aspect ratio and probably drop the top and bottom of the render buffer.

Alternatively you can drop a viewport into the scene, add VROrigin and subnodes to that viewport and turn ARVR on on that viewport instead of turning it on in the main viewport. Now you're free to add a 3rd person camera to your main scene and it will render that camera to screen while it will render your ARVRCamera to the HMD.

The controllers should work however this is the last thing I added and I haven't been able to test it with an actual Vive myself (friend of mines who's hardware I was using to test moved interstate and while I can borrow a Vive from another friend, it won't run on my laptop). So it is very possible there is a typo preventing the actual update of the location. The problem probably relates to this line of code:
https://github.com/BastiaanOlij/godot/blob/ar_vr_server/scene/3d/arvr_nodes.cpp#L95

In the original implementation I was only creating device ids for controllers, now I'm creating ids for all types but they are unique per type, so you'll have controller 1, base station 1, anchor point 1, etc. That was one of the latest changes I made so I probably messed it up. I'll add some logging in the next iteration so we can see how it maps the controller ids in Godot to the controllers in OpenVR.

beniwtv commented:
Hmm, looking at that line, it seems the same line is used in the "ARVRController::get_controller_name(void)" as well, so I tried calling get_controller_name() from my GDScript and it does indeed return the name of the controller.

I am wondering if "int joy_id = tracker->get_joy_id();" below is failing - if I call get_connected_joypads( ) from the Input class in my GDScript - I get an empty array. Is Godot supposed to detect the Vive controllers in it's input system? I ask only because a few lines down it seems to be using the Input system to detect button presses.

BastiaanOlij commented:
Yup, when the OpenVR module detects a vive controller is turned on it creates a tracker (for recording positioning info) and a joystick entry (for buttons, triggers, etc) for that controller.

The ARVRController is simply a convenient way to have Godot do some heavy lifting for you. 
I would have expected get_connected_joypads to return an entry for each controller.

Have a look at what ARVRServer.get_tracker_count returns?

beniwtv commented:
So I did some more investigation:

- get_tracker_count() returns 4. Two of type 1 (controller), two of type 2 (base station).
- get_connected_joypads( ) also does in fact return two controllers: ID 0 => vr_controller_vive_1_5_3 and ID 1 =>  vr_controller_vive_1_5_4; I just had it at the wrong spot, doh!

I looped the trackers to get some debug:
```

Name: lh_basestation_vive_1
Type: 2
Tracks position: True
Tracks orientation: True
Joy ID: -1

Name: lh_basestation_vive_2
Type: 2
Tracks position: True
Tracks orientation: True
Joy ID: -1

Name: vr_controller_vive_1_5_3 
Type: 1
Tracks position: True
Tracks orientation: True
Joy ID: 0

Name: vr_controller_vive_1_5_4
Type: 1
Tracks position: True
Tracks orientation: True
Joy ID: 1

```

So it does in fact seem to register both controllers. However, I noticed something odd. I connected the button_press signal to both ARVRControllers in the scene, and oddly enough, I can only get the controller that has the ID '2' in the inspector to register button presses. And it's always the same Vive controller, no matter what ARVRController I assign that ID to.

So it seems your hunch is right about some IDs being messed up. So I debugged some more, and found there is a wrong assumption:

https://github.com/BastiaanOlij/godot/blob/ar_vr_server/scene/3d/arvr_nodes.cpp#L102

As seen above, joystick ids seem to start from 0, checking "!= 0" effectively prevented one controller from registering key presses. I hacked it to "> -1" (probably not the right solution), and both controllers started having button events.

However, the controllers still do not track. I'll investigate some more in the coming days.

BastiaanOlij commented:
@beniwtv , hehe I already see whats wrong (besides the incorrect check for 0 instead of -1), I've forgotten to actually get the transform and apply it to the node. 

I think the line that is missing could be added on line 101 right after the else and probably should be:
`set_transform(tracker->get_transform(true)); // copy transform from tracker into node`

For debugging it may also be a good idea to export ARVRPositionalTracker::get_tracker_id to GDScript, then that little listing you just showed (nice one btw) can show the assigned tracker id as well. This should be unique per tracker type. So your first base station gets id 1, your second base station id 2, your first controller id 1 and your second controller id 2 (and the same for anchor points in AR). It are these IDs that are mapped to the controller nodes, so AVRController with controller_id = 1 links up to the tracker which is a controller with tracker id 1. 

I'm at work now but I'll clean it up tonight.

BastiaanOlij commented:
rebased, clang format issue on arvr server has already been resolved

BastiaanOlij commented:
Hey @akien, travis falling over nullptr is that another Cxx issue with mingw ?

beniwtv commented:
@BastiaanOlij Did a fresh clone/pull/compile and I can report everything seems to be working now nice!
(Apart from OpenGL not optimized under Linux for VR, but that's not Godot's fault)

BastiaanOlij commented:
Thanks @beniwtv , your help has been great!

beniwtv commented:
Oh cool, going to give this a try and report back. Wanted to test OpenHMD too but these weeks have been pretty busy :(

BastiaanOlij commented:
Same here Ben, it's been a bit crazy at work lately and it doesn't look like things will improve much before September. Can't wait to get back into this now that I have a machine that actually works with a Vive and hopefully will be able to borrow one from my friend soon.

svntax commented:
I managed to set up and test the example project on Windows 10 with an Oculus Rift + Oculus Touch controllers, and the tracking works great! But unfortunately I'm getting problems with detecting controller input.

For the button_pressed signal in ARVRController, the only buttons that seem to be detected on the Touch controllers are the grip button and the A, B, X, and Y buttons. I added a script to each ARVRController node to print out the values being passed into the signal and get the following numbers:
- 2 for the grip button
- 7 for the A and X buttons
- 1 for the B and Y buttons

I couldn't figure out how to get the thumbsticks working either, so in the end, the thumbsticks and the trigger button don't seem to be doing anything for me.

For the button_release signal, nothing seems to be detected at all. I added a function in the same script as before, connected it to the button_release signal, and tried pressing every single button on the Touch controllers (the 2 menu buttons bring up the SteamVR menu and the Oculus menu, but that's fine since they're already reserved), but the function never runs. Even just adding a print statement in the function doesn't output anything.

I'm not sure what's causing these problems, but everything else works fine though. The headset and the controllers' positions and orientations are tracked correctly, so this is already looking good.

BastiaanOlij commented:
Hopefully I'll soon have access to a vive again and actually debug my own code :)

I'm guessing the button_release signal not working is something wrong in code.

The triggers are likely analogue axis, not buttons, so you have to read out their positions. We could always add functionality that they also work as buttons when their value get above a certain value probably with some hysteresis logic to prevent the buttons from going on/off to erratically. 

Cool though that its all working with Oculus devices

beniwtv commented:
Did test the rebase; seems to be working fine for me. I haven't checked all buttons / axis', and signals for the buttons, will do that and report back.

beniwtv commented:
So here are my findings so far:

button_pressed signal: Working
button_release: Triggers constantly, always on button 0. Can't get it to trigger for other buttons
axis: Always returns 0 on all axis when calling the get_joystick_axis( int axis ) function

BastiaanOlij commented:
Rebased to latest master

BastiaanOlij commented:
@beniwtv   Found the bug with retrieving the button and axis values. As this is a bug in the ARVR server itself you'll need to apply #11280 to fix :)

BastiaanOlij committed:
Made a few tweaks to the interface

BastiaanOlij commented:
Okay, rebased to latest master. Note that you need #11712 to fix a compile error on Windows that snuck into master.

Small change in detecting any devices with the name camera in it so that it doesn't think its a controller. The rifts camera was picked up as a controller, oops :)

beniwtv commented:
@BastiaanOlij Finally got some time and wanted to test this, but I ran into compile errors :(

1) There's an env_base.Append in the Scons file - the CI above runs into the same issue.
2) Once I fixed that, I ran into another issue:

```
[ 30%] In file included from core/vector.h:39:0,
                 from core/ustring.h:35,
                 from ./core/math/math_2d.h:34,
                 from ./core/math/camera_matrix.h:33,
                 from ./servers/arvr/arvr_interface.h:33,
                 from modules/mobile_vr/mobile_interface.h:33,
                 from modules/mobile_vr/register_types.cpp:32:
core/reference.h: In instantiation of 'void Ref<T>::instance() [with T = MobileVRInterface]':
modules/mobile_vr/register_types.cpp:38:21:   required from here
core/os/memory.h:88:42: error: invalid new-expression of abstract class type 'MobileVRInterface'
 #define memnew(m_class) _post_initialize(new ("") m_class)
core/reference.h:268:7: note: in expansion of macro 'memnew'
   ref(memnew(T));
       ^~~~~~
In file included from modules/mobile_vr/register_types.cpp:32:0:
modules/mobile_vr/mobile_interface.h:51:7: note:   because the following virtual functions are pure within 'MobileVRInterface':
 class MobileVRInterface : public ARVRInterface {
       ^~~~~~~~~~~~~~~~~
In file included from modules/mobile_vr/mobile_interface.h:33:0,
                 from modules/mobile_vr/register_types.cpp:32:
./servers/arvr/arvr_interface.h:85:14: note: 	virtual int ARVRInterface::get_capabilities() const
  virtual int get_capabilities() const = 0;
              ^~~~~~~~~~~~~~~~
[ 30%] In file included from core/vector.h:39:0,
                 from core/ustring.h:35,
                 from ./core/math/math_2d.h:34,
                 from ./core/math/camera_matrix.h:33,
                 from ./servers/arvr/arvr_interface.h:33,
                 from modules/mobile_vr/mobile_interface.h:33,
                 from modules/mobile_vr/register_types.cpp:32:
core/class_db.h: In instantiation of 'static Object* ClassDB::creator() [with T = MobileVRInterface]':
core/class_db.h:170:20:   required from 'static void ClassDB::register_class() [with T = MobileVRInterface]'
modules/mobile_vr/register_types.cpp:35:45:   required from here
core/os/memory.h:88:42: error: invalid new-expression of abstract class type 'MobileVRInterface'
 #define memnew(m_class) _post_initialize(new ("") m_class)
core/class_db.h:137:10: note: in expansion of macro 'memnew'
   return memnew(T);
          ^~~~~~
```

Compiling from your fork works though, so it may be something that needs rebasing. Anyhow, on your branch (with the verified fix), I still can't get axis to work at all (tried axis numbers 0-16), all return just 0; and there is a continous button release event for button 0, that is emitted every tick.

BastiaanOlij commented:
Hey Ben,

Keep in mind that there are a few PRs on top of eachother because I'm still working through a few renames and polish ups that haven't been merged into master yet. I'm also not sure why you're getting compile issues on mobilevr as that should not be part of this pr at all. 

So yeah, best to work off of my fork to make sure you have the correct combination of commits :)

I've been stuck the last few days trying to get the sky background to work in stereoscopic rendering, its been doing my head in, but on the up side, my rift is working beautifully with openvr:) don't have controller though

BastiaanOlij commented:
Ah! The mobile VR interface was merged into master. I hadn't noticed that. That hasn't been updated to match some of the latest changes. I better polish that up tonight :)

With the default code it will pick the first applicable interface it can find so you'll need to change your init code to pick OpenVR over mobile VR. I need to think that through a little more

BastiaanOlij commented:
OK everyone, I'm closing this one :)

Pending testing and some debugging this PR #11937 implements our GDNative framework for ARVR drivers. Last night I had the HMD tracking working fine and while I haven't tested it yet controller handling should work now as well.

You can find the GDNative based OpenVR driver for Godot here:
https://github.com/BastiaanOlij/godot_openvr
This also contains a test project. I'll be polishing this up more over the next couple of days but it has feature parity with this PR.
